### ISSUE TITLE
etcd fails to register with discovery service

### ISSUE THREAD

**Issue by crawford** [coreos/etcd#3171]
This was etcd 2.0.13 on CoreOS 745.1.0.

```
$ journalctl -t etcd2

-- Logs begin at Tue 2015-07-21 03:19:15 UTC, end at Tue 2015-07-21 03:30:30 UTC. --
Jul 21 03:19:47 ip-10-0-0-22.ec2.internal etcd2[684]: 2015/07/21 03:19:47 etcd: listening for peers on http://10.0.0.22:2380
Jul 21 03:19:47 ip-10-0-0-22.ec2.internal etcd2[684]: 2015/07/21 03:19:47 etcd: listening for client requests on http://0.0.0.0:2379
Jul 21 03:19:47 ip-10-0-0-22.ec2.internal etcd2[684]: 2015/07/21 03:19:47 etcdserver: datadir is valid for the 2.0.1 format
Jul 21 03:19:53 ip-10-0-0-22.ec2.internal etcd2[684]: 2015/07/21 03:19:53 etcd: stopping listening for client requests on http://0.0.0.0:2379
Jul 21 03:19:53 ip-10-0-0-22.ec2.internal etcd2[684]: 2015/07/21 03:19:53 etcd: stopping listening for peers on http://10.0.0.22:2380
Jul 21 03:19:53 ip-10-0-0-22.ec2.internal etcd2[684]: 2015/07/21 03:19:53 etcd: context deadline exceeded
Jul 21 03:20:03 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:03 etcd: listening for peers on http://10.0.0.22:2380
Jul 21 03:20:03 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:03 etcd: listening for client requests on http://0.0.0.0:2379
Jul 21 03:20:03 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:03 etcdserver: datadir is valid for the 2.0.1 format
Jul 21 03:20:04 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:04 etcd: stopping listening for client requests on http://0.0.0.0:2379
Jul 21 03:20:04 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:04 etcd: stopping listening for peers on http://10.0.0.22:2380
Jul 21 03:20:04 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:04 member "795d4b9be5a74d6fb970f745359fb069" has previously registered with discovery service token (https://discovery.e
Jul 21 03:20:04 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:04 But etcd could not find vaild cluster configuration in the given data dir (/var/lib/etcd2).
Jul 21 03:20:04 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:04 Please check the given data dir path if the previous bootstrap succeeded
Jul 21 03:20:04 ip-10-0-0-22.ec2.internal etcd2[756]: 2015/07/21 03:20:04 or use a new discovery token if the previous bootstrap failed.
```

**Comment by yichengq** [comment]
1. It failed at the first time because discovery service told it that its proposal timed out. But the proposal is actually committed, and the same proposal after restart is rejected for cluster safety. 
2. We cannot totally avoid this problem because discovery service is based on etcd, which cannot guarantee to report the result whether each proposal is committed.
3. It should happen much less. current discovery service uses 2.0 and increases the probability of leader election(18 times in 150min), which makes proposal easier to time out. We plan to move discovery service to 2.1 soon.
4. The error message is hard to understand, and it should be improved.

**Comment by xiang90** [comment]
@crawford It seems that @yichengq has an answer for you. I am closing this due to low activity. Let us know if you have any other concern/questions.

**Comment by MiguelPF** [comment]
Hello the same is happening to me, can anyone suggest a fix for this, I am not able to make it work

thanks in advance

**Comment by yichengq** [comment]
@MiguelPF Using a new discovery token could solve this case.

And etcd 2.1.2+ supports better discovery error handling, and this should happen rarely.

**Comment by hectorj2f** [comment]
@yichengq I just reproduced it with etcd v2.1.2 on a CoreOS 681.2.0 :/

**Comment by nixlike** [comment]
Hi!

The problem is still there , the below log from running 3 node vagrant cluster 

ID=coreos
VERSION=884.0.0
VERSION_ID=884.0.0
BUILD_ID=2015-12-03-0237
PRETTY_NAME="CoreOS 884.0.0"

```
Dec 09 18:45:53 core-01 etcd2[3946]: etcd Version: 2.2.2
Dec 09 18:45:53 core-01 etcd2[3946]: Git SHA: b4bddf6
Dec 09 18:45:53 core-01 etcd2[3946]: Go Version: go1.4.3
Dec 09 18:45:53 core-01 etcd2[3946]: Go OS/Arch: linux/amd64
Dec 09 18:45:53 core-01 etcd2[3946]: setting maximum number of CPUs to 1, total number of available CPUs is 1
Dec 09 18:45:53 core-01 etcd2[3946]: the server is already initialized as member before, starting as etcd member...
Dec 09 18:45:53 core-01 etcd2[3946]: listening for peers on http://172.17.8.101:2380
Dec 09 18:45:53 core-01 etcd2[3946]: listening for peers on http://172.17.8.101:7001
Dec 09 18:45:53 core-01 etcd2[3946]: listening for client requests on http://0.0.0.0:2379
Dec 09 18:45:53 core-01 etcd2[3946]: listening for client requests on http://0.0.0.0:4001
Dec 09 18:45:55 core-01 etcd2[3946]: stopping listening for client requests on http://0.0.0.0:4001
Dec 09 18:45:55 core-01 etcd2[3946]: stopping listening for client requests on http://0.0.0.0:2379
Dec 09 18:45:55 core-01 etcd2[3946]: stopping listening for peers on http://172.17.8.101:7001
Dec 09 18:45:55 core-01 etcd2[3946]: stopping listening for peers on http://172.17.8.101:2380
Dec 09 18:45:55 core-01 etcd2[3946]: member "61b5f84c96044bc59b2b27c8fce8e277" has previously registered with discovery servic
Dec 09 18:45:55 core-01 etcd2[3946]: But etcd could not find valid cluster configuration in the given data dir (/var/lib/etcd2
Dec 09 18:45:55 core-01 systemd[1]: etcd2.service: Main process exited, code=exited, status=1/FAILURE
Dec 09 18:45:55 core-01 systemd[1]: Failed to start etcd2.
Dec 09 18:45:55 core-01 systemd[1]: etcd2.service: Unit entered failed state.
Dec 09 18:45:55 core-01 systemd[1]: etcd2.service: Failed with result 'exit-code'.
lines 5912-5940/5940 (END)

```

**Comment by bweston92** [comment]
I've just got this error now on DigitalOcean. Tried multiple times now.

```
-- Unit etcd2.service has begun starting up.
Jan 16 21:44:49 alicia etcd2[1773]: recognized and used environment variable ETCD_ADVERTISE_CLIENT_URLS=http://46.101.88.185:2379
Jan 16 21:44:49 alicia etcd2[1773]: recognized and used environment variable ETCD_DATA_DIR=/var/lib/etcd2
Jan 16 21:44:49 alicia etcd2[1773]: recognized and used environment variable ETCD_DISCOVERY=https://discovery.etcd.io/b0ed396b0d4b3a81bbf6cfc7666079fd
Jan 16 21:44:49 alicia etcd2[1773]: recognized and used environment variable ETCD_INITIAL_ADVERTISE_PEER_URLS=http://10.131.0.187:2380
Jan 16 21:44:49 alicia etcd2[1773]: recognized and used environment variable ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379,http://0.0.0.0:4001
Jan 16 21:44:49 alicia etcd2[1773]: recognized and used environment variable ETCD_LISTEN_PEER_URLS=http://10.131.0.187:2380,http://10.131.0.187:7001
Jan 16 21:44:49 alicia etcd2[1773]: recognized and used environment variable ETCD_NAME=0abd079fff964fd8b43ca190bc71afbf
Jan 16 21:44:49 alicia etcd2[1773]: etcd Version: 2.2.0
Jan 16 21:44:49 alicia etcd2[1773]: Git SHA: e4561dd
Jan 16 21:44:49 alicia etcd2[1773]: Go Version: go1.4.2
Jan 16 21:44:49 alicia etcd2[1773]: Go OS/Arch: linux/amd64
Jan 16 21:44:49 alicia etcd2[1773]: setting maximum number of CPUs to 1, total number of available CPUs is 1
Jan 16 21:44:49 alicia etcd2[1773]: the server is already initialized as member before, starting as etcd member...
Jan 16 21:44:49 alicia etcd2[1773]: listening for peers on http://10.131.0.187:2380
Jan 16 21:44:49 alicia etcd2[1773]: listening for peers on http://10.131.0.187:7001
Jan 16 21:44:49 alicia etcd2[1773]: listening for client requests on http://0.0.0.0:2379
Jan 16 21:44:49 alicia etcd2[1773]: listening for client requests on http://0.0.0.0:4001
Jan 16 21:44:50 alicia etcd2[1773]: stopping listening for client requests on http://0.0.0.0:4001
Jan 16 21:44:50 alicia etcd2[1773]: stopping listening for client requests on http://0.0.0.0:2379
Jan 16 21:44:50 alicia etcd2[1773]: stopping listening for peers on http://10.131.0.187:7001
Jan 16 21:44:50 alicia etcd2[1773]: stopping listening for peers on http://10.131.0.187:2380
Jan 16 21:44:50 alicia etcd2[1773]: member "0abd079fff964fd8b43ca190bc71afbf" has previously registered with discovery service token (https://discovery.etcd.io/b0ed396b0d4b3a81bbf6cfc7666079fd).
Jan 16 21:44:50 alicia etcd2[1773]: But etcd could not find valid cluster configuration in the given data dir (/var/lib/etcd2).
Jan 16 21:44:50 alicia etcd2[1773]: Please check the given data dir path if the previous bootstrap succeeded
Jan 16 21:44:50 alicia etcd2[1773]: or use a new discovery token if the previous bootstrap failed.
Jan 16 21:44:50 alicia polkitd[1671]: Unregistered Authentication Agent for unix-process:1768:115670 (system bus name :1.55, object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale en_US.UTF-8) (disconnected from bus)
Jan 16 21:44:50 alicia systemd[1]: etcd2.service: Main process exited, code=exited, status=1/FAILURE
Jan 16 21:44:50 alicia systemd[1]: Failed to start etcd2.
-- Subject: Unit etcd2.service has failed
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit etcd2.service has failed.
--
-- The result is failed.
Jan 16 21:44:50 alicia systemd[1]: etcd2.service: Unit entered failed state.
Jan 16 21:44:50 alicia systemd[1]: etcd2.service: Failed with result 'exit-code'.
Jan 16 21:44:52 alicia systemd[1]: Started OpenSSH per-connection server daemon (212.156.88.46:48079).
-- Subject: Unit sshd@40-46.101.88.185:22-212.156.88.46:48079.service has finished start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
--
-- Unit sshd@40-46.101.88.185:22-212.156.88.46:48079.service has finished starting up.
--
-- The start-up result is done.
Jan 16 21:44:52 alicia sshd[1780]: Connection closed by 212.156.88.46 [preauth]
lines 1643-1704/1704 (END)
```

**Comment by ahosam** [comment]
same issue here to install BM CoreOS  

Sep 04 12:43:33 core-01 systemd[1]: Starting etcd2...
Sep 04 12:43:33 core-01 etcd2[7804]: recognized and used environment variable ETCD_ADVERTISE_CLIENT_URLS=http://192.168.1.120:2379
Sep 04 12:43:33 core-01 etcd2[7804]: recognized and used environment variable ETCD_DATA_DIR=/var/lib/etcd2
Sep 04 12:43:33 core-01 etcd2[7804]: recognized and used environment variable ETCD_DISCOVERY=https://discovery.etcd.io/d899c284cd13fa8fc2d66d36fdb3436c
Sep 04 12:43:33 core-01 etcd2[7804]: recognized and used environment variable ETCD_INITIAL_ADVERTISE_PEER_URLS=http://192.168.1.120:2380
Sep 04 12:43:33 core-01 etcd2[7804]: recognized and used environment variable ETCD_LISTEN_CLIENT_URLS=http://0.0.0.0:2379
Sep 04 12:43:33 core-01 etcd2[7804]: recognized and used environment variable ETCD_LISTEN_PEER_URLS=http://192.168.1.120:2380
Sep 04 12:43:33 core-01 etcd2[7804]: recognized and used environment variable ETCD_NAME=4c980d5ad99944e78bb04dc6cd4aebc2
Sep 04 12:43:33 core-01 etcd2[7804]: etcd Version: 2.3.2
Sep 04 12:43:33 core-01 etcd2[7804]: Git SHA: ce63f10
Sep 04 12:43:33 core-01 etcd2[7804]: Go Version: go1.5.4
Sep 04 12:43:33 core-01 etcd2[7804]: Go OS/Arch: linux/amd64
Sep 04 12:43:33 core-01 etcd2[7804]: setting maximum number of CPUs to 2, total number of available CPUs is 2
Sep 04 12:43:33 core-01 etcd2[7804]: listening for peers on http://192.168.1.120:2380
Sep 04 12:43:33 core-01 etcd2[7804]: listening for client requests on http://0.0.0.0:2379
Sep 04 12:43:35 core-01 etcd2[7804]: stopping listening for client requests on http://0.0.0.0:2379
Sep 04 12:43:35 core-01 etcd2[7804]: stopping listening for peers on http://192.168.1.120:2380
Sep 04 12:43:35 core-01 etcd2[7804]: member "4c980d5ad99944e78bb04dc6cd4aebc2" has previously registered with discovery service token (https://discovery.etcd.io/d899c284cd13fa8fc2d66d36fdb3436c).
Sep 04 12:43:35 core-01 etcd2[7804]: But etcd could not find valid cluster configuration in the given data dir (/var/lib/etcd2).
Sep 04 12:43:35 core-01 etcd2[7804]: Please check the given data dir path if the previous bootstrap succeeded
Sep 04 12:43:35 core-01 etcd2[7804]: or use a new discovery token if the previous bootstrap failed.
Sep 04 12:43:35 core-01 systemd[1]: etcd2.service: Main process exited, code=exited, status=1/FAILURE
Sep 04 12:43:35 core-01 systemd[1]: Failed to start etcd2.
Sep 04 12:43:35 core-01 systemd[1]: etcd2.service: Unit entered failed state.
Sep 04 12:43:35 core-01 systemd[1]: etcd2.service: Failed with result 'exit-code'.
